content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
++++
+title = "Automatic Static Memory Management in Bril"
+[extra]
+bio = """
+  Patrick LaFontaine is a senior undergraduate student interested in programming language design and implementation. He also enjoys backpacking and attempting to bake the perfect loaf of bread.
+"""
+[[extra.authors]]
+name = "Patrick LaFontaine"
++++
+
+## Motivation: Compiler-led Garbage Collection
+
+The [memory extension][memex] for the Bril programming language provides manually-managed memory for the allocation of arrays. Manual memory management, when the programmer is in charge of explicitly freeing their memory, provides full control to the programmer at the cost of exposing them to difficult-to-debug memory bugs. The common bugs are leaking memory by forgetting to call ```free```, attempting to ```free``` memory more than once, or attempting to use a memory location after it has been freed. A popular way to handle these issues is to have a run-time system in charge of deciding when to ```free``` memory. These run-time garbage collectors allow programmers to easily work with memory, but often at the cost of worse performance, poor latency with "stop the world" pauses or significant memory usage. One alternative to this is to put the compiler in charge of deciding when to free memory. This handles all of the memory management logic, at compile time, for the programmer, hence "Automatic Static Memory Management". The main downside of this approach is that this static analysis, like all compiler analyses, must be conservative which can limit the expressiveness of the programming language.
+
+## A Region-Based type extension for Bril
+
+This work, based on the [cyclone] language, seeks to bring region-based memory management to the Bril [memory extension][memex]. Each pointer to memory exists within the static scope of assigned memory called a region. Regions are hierarchical such that they can be nested within each other. Data that is allocated inside of a region can only be used within that region or inner-nested regions. When the scope of a region ends, all of the allocations that belong to that region are freed. In this way, the compiler statically rules out the previously stated, common memory bugs by enforcing when and for how long an allocation can be used for. For Bril, this comes in the form of an extension to the Bril Json representation and a static analysis tool that accepts this extension of Bril and converts it to standard Bril with manual memory management inserted into the program.
+
+## Extension Design
+
+The Type enum in the [bril_rs] library is modified as follows:
```

Comment:
I think it would be more helpful to just give the JSON representation instead of the Rust data type. People might not be familiar with the internals of your Rust library design, whereas using the JSON notation will match the Bril docs exactly. (And it will help clarify the separation between the abstract, implementation-independent language design and the Rust implementation.)

Reply:
Makes sense.